src/html5/parser/document.rs

```diff
@@ -692,6 +698,59 @@ impl DocumentHandle {
 
         Ok(())
     }
+
+    /// Perform a single query against the document.
+    /// If query search type is uninitialized, returns an error.
+    /// Otherwise, returns a vector of NodeIds that match the predicate in tree order.
+    pub fn query(&self, query: &Query) -> Result<Vec<NodeId>> {
+        if query.search_type == SearchType::Uninitialized {
+            return Err(Error::Query("Query predicate is uninitialized".to_owned()));
+        }
+
+        let mut found_ids = Vec::new();
+
+        let mut node_queue: Queue<NodeId> = queue![];
+        let root_id = self.get().get_root().id;
+        let _ = node_queue.add(root_id);
```

Comment:
In practice I'm not sure how often the `add/remove` methods fail on a queue, but it may be good to check these errors just for completion...

Reply:
no longer relevant, queue structure removed